### PR TITLE
NER example: fix metrics computation

### PR DESCRIPTION
**Pull request by WilliamTambellini**
## Description ##
NER example: fix metrics computation

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [X] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [X] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [X] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [X] Fix metrics computation

## Comments ##
- this change is backward compatible

### CODE REVIEW

**Comment by abhinavs95** [comment]
Thanks for the fix @WilliamTambellini 

@mxnet-label-bot add [Example, Python, pr-awaiting-review]

File: example/named_entity_recognition/src/metrics.py [hunk 1]
```diff
@@ -64,6 +65,8 @@ def classifer_metrics(label, pred):
         f1 = 0
     else:
         f1 = 2 * precision * recall / (precision + recall)
+
+    logging.log(logging.DEBUG, "Metrics results: precision=%f recall=%f f1=%f" % (precision, recall, f1))
```
Comment (szha) [inline example/named_entity_recognition/src/metrics.py]:
```suggestion
    logging.debug("Metrics results: precision=%f recall=%f f1=%f", precision, recall, f1)
```

File: example/named_entity_recognition/src/ner.py [hunk 2]
```diff
@@ -93,6 +93,7 @@ def build_vocab(nested_list):
     """
     # Build vocabulary
     word_counts = Counter(itertools.chain(*nested_list))
+    logging.getLogger().info("build_vocab: word_counts=%d" % (len(word_counts)))
```
Comment (szha) [inline example/named_entity_recognition/src/ner.py]:
```suggestion
    logging.info("build_vocab: word_counts=%d", len(word_counts))
```

File: example/named_entity_recognition/src/ner.py [hunk 3]
```diff
@@ -135,12 +137,14 @@ def build_iters(data_dir, max_records, train_fraction, batch_size, buckets=None)
 
     # Split into training and testing data
     idx=int(len(indexed_tokens)*train_fraction)
+    logging.getLogger().info("Preparing train/test datasets splitting at idx %d on total %d sentences using a batchsize of %d" % (idx, len(indexed_tokens), batch_size))
```
Comment (szha) [inline example/named_entity_recognition/src/ner.py]:
```suggestion
    logging.info("Preparing train/test datasets splitting at idx %d on total %d sentences using a batchsize of %d", idx, len(indexed_tokens), batch_size)
```

File: example/named_entity_recognition/src/ner.py [hunk 4]
```diff
@@ -135,12 +137,14 @@ def build_iters(data_dir, max_records, train_fraction, batch_size, buckets=None)
...
     X_token_train, X_char_train, Y_train = indexed_tokens[:idx], indexed_chars[:idx], indexed_entities[:idx]
     X_token_test, X_char_test, Y_test = indexed_tokens[idx:], indexed_chars[idx:], indexed_entities[idx:]
 
     # build iterators to feed batches to network
     train_iter = iterators.BucketNerIter(sentences=X_token_train, characters=X_char_train, label=Y_train,
                                          max_token_chars=5, batch_size=batch_size, buckets=buckets)
+    logging.getLogger().info("Creating the val_iter using %d sentences" % (len(X_token_test)))
```
Comment (szha) [inline example/named_entity_recognition/src/ner.py]:
```suggestion
    logging.info("Creating the val_iter using %d sentences", len(X_token_test))
```

File: example/named_entity_recognition/src/ner.py [hunk 5]
```diff
@@ -205,6 +209,8 @@ def sym_gen(seq_len):
 def train(train_iter, val_iter):
     import metrics
     devs = mx.cpu() if args.gpus is None or args.gpus is '' else [mx.gpu(int(i)) for i in args.gpus.split(',')]
+    logging.getLogger().info("train on device %s using optimizer %s at learningrate %f for %d epochs using %d records: lstm_state_size=%d ..."
```
Comment (szha) [inline example/named_entity_recognition/src/ner.py]:
```suggestion
    logging.info("train on device %s using optimizer %s at learningrate %f for %d epochs using %d records: lstm_state_size=%d ...", 
                 devs, args.optimizer, args.lr, args.num_epochs, args.max_records, args.lstm_state_size)
```

File: example/named_entity_recognition/src/ner.py [hunk 6]
```diff
@@ -205,6 +209,8 @@ def sym_gen(seq_len):
 def train(train_iter, val_iter):
     import metrics
     devs = mx.cpu() if args.gpus is None or args.gpus is '' else [mx.gpu(int(i)) for i in args.gpus.split(',')]
+    logging.getLogger().info("train on device %s using optimizer %s at learningrate %f for %d epochs using %d records: lstm_state_size=%d ..." 
+          % (devs, args.optimizer, args.lr, args.num_epochs, args.max_records, args.lstm_state_size))
```
Comment (szha) [inline example/named_entity_recognition/src/ner.py]:
```suggestion
```

File: example/named_entity_recognition/src/ner.py [hunk 7]
```diff
@@ -225,6 +231,8 @@ def train(train_iter, val_iter):
     train_iter, val_iter, word_to_index, char_to_index, entity_to_index = build_iters(args.data_dir, args.max_records,
                                                                      args.train_fraction, args.batch_size, args.buckets)
 
+    logging.getLogger().info("validation iterator: %s" % (val_iter))
```
Comment (szha) [inline example/named_entity_recognition/src/ner.py]:
```suggestion
    logging.info("validation iterator: %s, val_iter)
```

**Comment by vandanavk** [comment]
@piyushghai @abhinavs95 @karan6181 for review

**Comment by szha** [comment]
@WilliamTambellini seems that there's one more to fix for the lazy logging (the open comment).

**Comment by WilliamTambellini** [comment]
> @WilliamTambellini seems that there's one more to fix for the lazy logging (the open comment).

done

**Comment by WilliamTambellini** [comment]
hum the unix-gpu job failed with
Failed:   1
Warnings: 4
Skipped:  2
Error: Test failures
Execution halted
Makefile:677: recipe for target 'rpkgtest' failed
make: *** [rpkgtest] Error 1

build.py: 2019-06-20 17:34:39,235Z CRITICAL Execution of ['/work/runtime_functions.sh', 'unittest_ubuntu_gpu_R'] failed with status: 2

Is that job supposed to pass ?

**Comment by szha** [comment]
@WilliamTambellini I triggered these builds again.

**Comment by WilliamTambellini** [comment]
@szha  tks, it's better but now unix-cpu failed.

**Comment by karan6181** [comment]
@WilliamTambellini is this PR good to go for merge? I see all CI test are passing.
